src/cassette.jl

```diff
@@ -141,6 +143,143 @@ function replacefunc(f::Function, d::AbstractDict)
     ctx = GraftCtx(metadata=d)
     return Cassette.recurse(ctx, f)
 end
+            
+#------------------------------------------------------------------------------------           
+            
+Cassette.@context TypeCtx
+            
+"""   TypeCtx
+
+creates a MetaDiGraph tracking the types of args and ret values throughout a script
+
+"""
+TypeCtx
+            
+            
+
+"""   FCollector(depth::Int,frame::function,data::FCollector)
+
+struct to collect all the functions called throughout a script
+        
+"""
+mutable struct FCollector{I,F,C}
+    depth::I
+    frame::F
+    data::Vector{C}
+end
+            
+
+
+"""   FCollector(depth::Int,frame::Frame)
+
+this is an initialization funtion for the FCollector
+
+"""
+function FCollector(d::Int, f)
+    FCollector(d, f, FCollector[])
+end
+
+""" Frame(func, args, ret, subtrace)
+
+a structure to hold metadata for recursive type information for every function called
+            
+"""
+mutable struct Frame{F,T,U}
+    func::F
+    args::T
+    ret::U
+end
+            
+# add boilerplate for functionality
+function Cassette.overdub(ctx::TypeCtx, f, args...)
+    c = FCollector(ctx.metadata.depth-1, Frame(f, args, Any))
+    push!(ctx.metadata.data, c)
+    if c.depth > 0 && Cassette.canrecurse(ctx, f, args...)
+        newctx = Cassette.similarcontext(ctx, metadata = c)
+        z = Cassette.recurse(newctx, f, args...)
+        c.frame.ret = typeof(z)
+        return z
+    else
+        z = Cassette.fallback(ctx, f, args...)
+        c.frame.ret = typeof(z)
+        return z
+    end
+end
+
+# limit the stacktrace in terms of which to recurse on
+Cassette.canrecurse(ctx::TypeCtx,::typeof(Base.vect), args...) = false
+Cassette.canrecurse(ctx::TypeCtx,::typeof(FCollector)) = false
+Cassette.canrecurse(ctx::TypeCtx,::typeof(Frame)) = false
+
+# this is a function that strips all using statements from a raw text file
+# reasons to do so
+# 1) we no longer include libraries we dont have too so semanticmodels can remain small
+# 2) prevents errors in the include-cassette design which doesn allow for usings
+function extractdeps(filename::String)
+    f = open(filename)
+    newf = open("new$filename","w") 
+    for line in readlines(f)
+        if occursin("using",line)
+            ex = Meta.parse(line)
```

Comment:
If we mandate that the input codes are self contained modules, ie

```julia
module MyModule
function foo(args...)
    ...
end
...
end
```
then we can call `Meta.parse(read(f))` and get all the code at once.

Reply:
This doesn't break the semantics of Julia, because all scripts are run in the modeule `Main` anyway.

Reply:
This is a test case I made for this style of input.
![image](https://user-images.githubusercontent.com/41996797/53339834-6d08e600-38d5-11e9-9108-c40366b60ca6.png)


I think this needs to be `readlines`. Read returns 
![image](https://user-images.githubusercontent.com/41996797/53339247-cff97d80-38d3-11e9-997b-8fcbb2b7b63c.png)
and `parse` will not accept it
![image](https://user-images.githubusercontent.com/41996797/53339297-f4edf080-38d3-11e9-8363-22cb4aa7f2e4.png)
however when I tired `readlines` it returns 
![image](https://user-images.githubusercontent.com/41996797/53339343-19e26380-38d4-11e9-8b40-78cd4f7b4d6c.png) 

The other issue is that cassette will try to recuse on `using` and the `module` statements which it is unable to resolve
![image](https://user-images.githubusercontent.com/41996797/53339800-52367180-38d5-11e9-8829-edc85a9038af.png)

Julia allows for this
![image](https://user-images.githubusercontent.com/41996797/53340199-662ea300-38d6-11e9-8158-f2e94c0167f3.png)

We can try this instead
![image](https://user-images.githubusercontent.com/41996797/53340167-4dbe8880-38d6-11e9-8af6-cc49475264ca.png)

Reply:
I closed some of the earlier comments so we can focus on this thread. If we change the called function for the overdub the rest become irrelevant.

Reply:
![image](https://user-images.githubusercontent.com/41996797/53341484-a2173780-38d9-11e9-9b20-af857f31b992.png)
This is another option. I am not sure if one should be preferred for speed or allocation?

Reply:
the reason not to use `readlines` is that it pulls each line out and you can't actually parse Julia code line by line, because you can have 

```julia
function f(x)
   return x + 2
end
```
which is 1 expression but 3 lines. so you would have to use a `read(String, file)` to get the whole file as one string.

We definitely should not overdub the include call, because that would trace a bunch of julia internals stuff we don't care about, like loading the script and parsing it.

Instead we should read, parse, and eval the script (wrapped in a module) and then call the module's main method.

See below for example using `Base.Meta.parse` and `eval`.